rfcs/0042-bitswap.md

```diff
@@ -0,0 +1,268 @@
+
+The daemon will have a new subsystem which will track the creation/destruction of resources, and perform the necessary caching protocols. Tracking the destruction of resources will be done using a GC finalizer. We cannot perform the "remove resource" caching protocol from inside of a finalizer, so the finalizer will merely mark removed resources for the subsystem to later remove from the Bitswap block cache when the Async scheduler schedules the subsystem to run.
+
+**NB:** _This proposed model of how the subsystem will track resources in relation to Bitswap block CIDs does not mesh well with the world in which we store ledgers (which are trees that will likely share blocks across various roots). We will need to layer a refcounting solution on top of this to support that use case, when the time comes._
+
+```ocaml
+module Bitswap = struct
+  type cid = <<elided>>
+  type block = <<elided>>
+  type block_tree = Bitswap.block * Bitswap.block list
+
+  module type Bitswap_advertisable_intf = sig
+    (* we use the bin_io interface here to serialize/deserialize the binary data stored in blocks *)
+    type t [@@deriving bin_io]
+
+    type create_args
+
+    val create : create_args -> t
+
+    (* mapping from an object to a mina cid *)
+    val id : t -> string 
+
+    val project_bitswap_blocks : t -> block_tree
+  end
+
+  type 'a advertisable = (module Bitswap_advertisable_intf with type t = 'a)
+
+  module Sync_state = struct
+    (* global sync state *)
+    let sync_table = String.Table.create ()
+
+    (* runs a continuous background task to synchronize resources tracked in the daemon with the Bitswap cache *)
+    let run_sync_task () =
+
+    (* track a new resource in global state *)
+    let track ((module Resource) : 'a advertisable) (resource : 'a) =
+      ...
+
+    (* mark a resource currently tracked in global state to be processed for collection in a future async cycle *)
+    let mark_for_collection ((module Resource) : 'a advertisable) (resource : 'a) =
+      ...
+  end
+
+  module Bitswap_advertised (Resource : Bitswap_advertisable_intf) = struct
+    type t = Resource.t
+
+    let create init_args =
+      let resource = Resource.create init_args in
+      Sync_state.add (module Resource) x ;
+      Gc.Expert.attach_finalizer x (fun _ -> Sync_state.mark_for_removal (module Resource) x) ;
+      x
+  end
+end
+```
+
+### Libp2p Helper IPC Modifications
+
+New IPC messages are required in order to support the daemon requesting resources stored in Bitswap blocks from the helper process. When the daemon needs to acquire a resource that is stored in Bitswap, it will send a message to the libp2p helper process asking it to download all the blocks associated with that resource (identified by the root Bitswap block CID). The helper process will begin the process of downloading all Bitswap blocks in the tree pointed to by the root Bitswap block, and will inform the helper upon success or failure of that operation.
```

Comment:
will there be some timeout in the daemon for the response?

Reply:
I was imagining the timeout would exist on the libp2p_helper side, and that the libp2p_helper would send a `ResourceDownloadError(root_cid: CID, error: String)` message if this timeout occurred.

Reply:
I'm updating the RFC to make this clear.